modules/mono/glue/GodotSharp/GodotSharp/Core/Variant.cs

```diff
@@ -109,16 +109,54 @@ public void Dispose()
 
     public override string ToString() => AsString();
 
-    public object? Obj
-    {
-        get
+    public object? Obj =>
+        _obj ??= NativeVar.DangerousSelfRef.Type switch
         {
-            if (_obj == null)
-                _obj = Marshaling.ConvertVariantToManagedObject((godot_variant)NativeVar);
-
-            return _obj;
-        }
-    }
+            Variant.Type.Bool => AsBool(),
+            Variant.Type.Int => AsInt64(),
+#if REAL_T_IS_DOUBLE
+            Variant.Type.Float => AsDouble(),
+#else
+            Variant.Type.Float => AsSingle(),
+#endif
+            Variant.Type.String => AsString(),
+            Variant.Type.Vector2 => AsVector2(),
+            Variant.Type.Vector2i => AsVector2i(),
+            Variant.Type.Rect2 => AsRect2(),
+            Variant.Type.Rect2i => AsRect2i(),
+            Variant.Type.Vector3 => AsVector3(),
+            Variant.Type.Vector3i => AsVector3i(),
+            Variant.Type.Transform2d => AsTransform2D(),
+            Variant.Type.Vector4 => AsVector4(),
+            Variant.Type.Vector4i => AsVector4i(),
+            Variant.Type.Plane => AsPlane(),
+            Variant.Type.Quaternion => AsQuaternion(),
+            Variant.Type.Aabb => AsAABB(),
+            Variant.Type.Basis => AsBasis(),
+            Variant.Type.Transform3d => AsTransform3D(),
+            Variant.Type.Projection => AsProjection(),
+            Variant.Type.Color => AsColor(),
+            Variant.Type.StringName => AsStringName(),
+            Variant.Type.NodePath => AsNodePath(),
+            Variant.Type.Rid => AsRID(),
+            Variant.Type.Object => AsGodotObject(),
+            Variant.Type.Callable => AsCallable(),
+            Variant.Type.Signal => AsSignalInfo(),
+            Variant.Type.Dictionary => AsGodotDictionary(),
+            Variant.Type.Array => AsGodotArray(),
+            Variant.Type.PackedByteArray => AsByteArray(),
+            Variant.Type.PackedInt32Array => AsInt32Array(),
+            Variant.Type.PackedInt64Array => AsInt64Array(),
+            Variant.Type.PackedFloat32Array => AsFloat32Array(),
+            Variant.Type.PackedFloat64Array => AsFloat64Array(),
+            Variant.Type.PackedStringArray => AsStringArray(),
+            Variant.Type.PackedVector2Array => AsVector2Array(),
+            Variant.Type.PackedVector3Array => AsVector3Array(),
+            Variant.Type.PackedColorArray => AsColorArray(),
+            Type.Nil => null,
+            Type.Max => null,
+            _ => null
```

Comment:
I don't think it's possible for a Variant to be of type `Max` or anything outside of the cases of the enum, so should it throw instead?